internal/controller/managedcluster_controller.go

```diff
@@ -110,60 +116,61 @@ func (r *ManagedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	return r.Update(ctx, managedCluster)
 }
 
-func (r *ManagedClusterReconciler) setStatusFromClusterStatus(ctx context.Context, managedCluster *hmc.ManagedCluster) (requeue bool, _ error) {
+func (r *ManagedClusterReconciler) setStatusFromChildObjects(ctx context.Context, managedCluster *hmc.ManagedCluster, resourceID schema.GroupVersionResource, requeueIfNoObjects bool) (requeue bool, _ error) {
 	l := ctrl.LoggerFrom(ctx)
 
-	resourceID := schema.GroupVersionResource{
-		Group:    "cluster.x-k8s.io",
-		Version:  "v1beta1",
-		Resource: "clusters",
-	}
-
 	list, err := r.DynamicClient.Resource(resourceID).Namespace(managedCluster.Namespace).List(ctx, metav1.ListOptions{
 		LabelSelector: labels.SelectorFromSet(map[string]string{hmc.FluxHelmChartNameKey: managedCluster.Name}).String(),
 	})
 
 	if apierrors.IsNotFound(err) || len(list.Items) == 0 {
-		l.Info("Clusters not found, ignoring since object must be deleted or not yet created")
-		return true, nil
+		message := resourceID.Resource + " not found, ignoring since object must be deleted or not yet created"
+		if requeueIfNoObjects {
+			l.Info(message)
+			return true, nil
+		}
+		return false, errors.New(message)
 	}
 
 	if err != nil {
-		return true, fmt.Errorf("failed to get cluster information for managedCluster %s in namespace: %s: %w",
-			managedCluster.Namespace, managedCluster.Name, err)
-	}
-	conditions, found, err := unstructured.NestedSlice(list.Items[0].Object, "status", "conditions")
-	if err != nil {
-		return true, fmt.Errorf("failed to get cluster information for managedCluster %s in namespace: %s: %w",
-			managedCluster.Namespace, managedCluster.Name, err)
-	}
-	if !found {
-		return true, fmt.Errorf("failed to get cluster information for managedCluster %s in namespace: %s: status.conditions not found",
-			managedCluster.Namespace, managedCluster.Name)
+		return true, fmt.Errorf("failed to get %s information for managedCluster %s in namespace: %s: %w",
+			resourceID.Resource, managedCluster.Namespace, managedCluster.Name, err)
 	}
 
 	allConditionsComplete := true
-	for _, condition := range conditions {
-		conditionMap, ok := condition.(map[string]any)
-		if !ok {
-			return true, fmt.Errorf("failed to cast condition to map[string]any for managedCluster: %s in namespace: %s: %w",
-				managedCluster.Namespace, managedCluster.Name, err)
+	for _, item := range list.Items {
```

Comment:
In current way it's a mess of conditions from different places with no relation to those places. I think we should cherry pick only relevant conditions and modify them to keep connection to the objects those conditions are collected from